File: leetcode/go/73.go
package main

func setZeroes(matrix [][]int) {
	firstRowZero := false
	m, n := len(matrix), len(matrix[0])
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if i == 0 && matrix[0][j] == 0 {
				firstRowZero = true
			} else if matrix[i][j] == 0 {
				matrix[i][0] = -1
				matrix[0][j] = -1
			}
		}
	}

	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			if matrix[i][0] == -1 || matrix[0][j] == -1 {
				matrix[i][j] = 0
			}
		}
	}
	if matrix[0][0] == -1 {
		matrix[0][0] = 0
		for i := 0; i < m; i++ {
			matrix[i][0] = 0
		}
	}
	if firstRowZero {
		for j := 0; j < n; j++ {
			matrix[0][j] = 0
		}
	}
}

File: leetcode/cpp/73.cpp

#include <bits/stdc++.h>
using namespace std;
class Solution1
{
    // brute force approach, mark all the required rows and cols
    void markRow(vector<vector<int>> matrix, int row_index)
    {
        int cols = matrix[0].size();
        for (int j = 0; j < cols; j++)
        {
            if (matrix[row_index][j] == 1)
            {
                matrix[row_index][j] = -1;
            }
        }
    }
    void markCol(vector<vector<int>> matrix, int col_index)
    {
        int rows = matrix.size();
        for (int i = 0; i < rows; i++)
        {
            if (matrix[i][col_index] == 1)
            {
                matrix[i][col_index] = -1;
            }
        }
    }

public:
    void setZeroes(vector<vector<int>> &matrix)
    {
        int n = matrix.size(), m = matrix[0].size();
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                markRow(matrix, i);
                markCol(matrix, j);
            }
        }

        // now making all -1 marked elements to 0 in a single-pass
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                if (matrix[i][j] == -1)
                {
                    matrix[i][j] = 0;
                }
            }
        }
        return;
        // Time complexity = m*n(m+n) + (m*n)
    }
};
int main()
{
}